ls1intum/Artemis
# [Feature] Grading System - Validation

### Checklist

### Motivation and Context


This PR introduces better client-side validation for the grade key create/edit page. The user will now get descriptive warnings when the grade key entered isn't following the correct format.

### Description

I've introduced a function that continuously checks for the validity of the grade steps and if something is incorrect, it disables the Save button and displays a descriptive warning to the user.

### Steps for Testing


1. Log in to Artemis as an instructor
2. Navigate to Course Administration -> Grade Key
3. Try entering invalid grade steps and check that appropriate warnings get displayed and the 'Save' button is disabled
a) No grade steps present
b) Empty fields (anywhere)
c) Grade percentages <0 and >100
d) Grade steps where Min % > Max %
e) Bonus points (must switch to 'Bonus' type) <0
f) Grade steps where the Max % and Min % of adjacent grade steps don't overlap
g) First grade step doesn't start with 0 (Min %)
h) Last grade step doesn't end with 100 (Max %)
4. In general go wild and try breaking the page every which way 😉

### Test Coverage



grading-system.component.ts: 94.34%

### Screenshots


Example scenario, when one of the grade steps is invalid:
![image](https://user-images.githubusercontent.com/63168603/120687636-dad7eb00-c4aa-11eb-8956-229e8af14c76.png)
